fpdf/line_break.py

```diff
@@ -0,0 +1,237 @@
+
+    def manual_break(self, justify=False):
+        return text_line(
+            self.fragments,
+            self.width,
+            self.number_of_spaces,
+            self.number_of_spaces and justify,
+        )
+
+    def automatic_break_possible(self):
+        return self.hyphen_break_hint is not None or self.space_break_hint is not None
+
+    def automatic_break(self, justify):
+        assert self.automatic_break_possible()
+        if self.hyphen_break_hint is not None and (
+            self.space_break_hint is None
+            or self.hyphen_break_hint.width > self.space_break_hint.width
+        ):
+            self._apply_automatic_hint(self.hyphen_break_hint)
+            self.add_character(
+                self.hyphen_break_hint.character_to_append,
+                self.hyphen_break_hint.character_to_append_width,
+                self.hyphen_break_hint.character_to_append_style,
+                self.hyphen_break_hint.character_to_append_underline,
+                self.hyphen_break_hint.original_fragment_index,
+                self.hyphen_break_hint.original_character_index,
+            )
+            return (
+                self.hyphen_break_hint.original_fragment_index,
+                self.hyphen_break_hint.original_character_index,
+                self.manual_break(justify),
+            )
+        self._apply_automatic_hint(self.space_break_hint)
+        return (
+            self.space_break_hint.original_fragment_index,
+            self.space_break_hint.original_character_index,
+            self.manual_break(justify),
+        )
+
+
+class MultiLineBreak:
+    def __init__(self, styled_text_fragments, size_by_style, justify=False):
+
+        self.styled_text_fragments = styled_text_fragments
+
+        self.size_by_style = size_by_style
+        self.justify = justify
+
+        self.fragment_index = 0
+        self.character_index = 0
+
+    def _get_character_width(self, character, style=""):
+        if character == SOFT_HYPHEN:
+            # HYPHEN is inserted instead of SOFT_HYPHEN
+            character = HYPHEN
+        return self.size_by_style(character, style)
+
+    def get_line_of_given_width(self, maximum_width):
+
+        if self.fragment_index == len(self.styled_text_fragments):
+            return None
+
+        current_line = CurrentLine()
+        while self.fragment_index < len(self.styled_text_fragments):
+
+            current_fragment = self.styled_text_fragments[self.fragment_index]
+
+            if self.character_index >= len(current_fragment.characters):
+                self.character_index = 0
+                self.fragment_index += 1
+                continue
+
+            character = current_fragment.characters[self.character_index]
+            character_width = self._get_character_width(
+                character, current_fragment.style
+            )
+
+            if character == NEWLINE:
+                return current_line.manual_break()
```

Comment:
When can this happen exactly?

Reply:
I will add testcases that explain why this is needed.